libportal/settings.c

```diff
@@ -0,0 +1,194 @@
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = xdp_settings_finalize;
+
+  /**
+   * XdpSettings::changed:
+   * @settings: the [class@Settings] object
+   * @namespace: the value namespace
+   * @key: the value key
+   * @value: the value
+   *
+   * Emitted when a setting value is changed externally.
+   */
+  signals[CHANGED] =
+    g_signal_new ("changed",
+                  G_TYPE_FROM_CLASS (object_class),
+                  G_SIGNAL_RUN_FIRST,
+                  0,
+                  NULL, NULL,
+                  NULL,
+                  G_TYPE_NONE, 3,
+                  G_TYPE_STRING,
+                  G_TYPE_STRING,
+                  G_TYPE_VARIANT);
+}
+
+static void
+xdp_settings_init (XdpSettings *settings)
+{
+}
+
+static void
+settings_changed (GDBusConnection *bus,
+		  const char *sender_name,
+		  const char *object_path,
+		  const char *interface_name,
+		  const char *signal_name,
+		  GVariant *parameters,
+		  gpointer data)
+{
+  const char *namespace = NULL;
+  const char *key = NULL;
+  g_autoptr(GVariant) value = NULL;
+  guint n_params;
+
+  XdpSettings *settings = data;
+
+  n_params = g_variant_n_children (parameters);
+  if (n_params != 3)
+    {
+      g_warning ("Incorrect number of parameters, expected 3, got %u", n_params);
+      return;
+    }
+
+  g_variant_get_child (parameters, 0, "&s", &namespace);
+  g_variant_get_child (parameters, 1, "&s", &key);
+  g_variant_get_child (parameters, 2, "v", &value);
+
+  g_signal_emit (settings, signals[CHANGED], 0, namespace, key, value);
+}
+
+/**
+ * xdp_settings_read_value:
+ * @settings: the [class@Settings] object.
+ * @namespace: the namespace of the value.
+ * @key: the key of the value.
+ * @cancellable: a GCancellable or NULL.
+ * @error: return location for error or NULL.
+ *
+ * Read a setting value within @namespace, with @key.
+ *
+ * Returns: (nullable) (transfer full): the value, or %NULL if not
+ * found. If @error is not NULL, then the error is returned.
+ */
+GVariant *
+xdp_settings_read_value (XdpSettings *settings, const char *namespace, const char *key, GCancellable *cancellable, GError **error)
+{
+  g_autoptr(GVariant) ret = NULL;
+  g_autoptr(GVariant) inner = NULL;
+
+  ret = g_dbus_connection_call_sync (settings->portal->bus,
+                                     PORTAL_BUS_NAME,
+                                     PORTAL_OBJECT_PATH,
+                                     SETTINGS_INTERFACE,
+                                     "ReadOne",
+                                     g_variant_new ("(ss)", namespace, key),
+                                     NULL,
+                                     G_DBUS_CALL_FLAGS_NONE,
+                                     -1,
+                                     cancellable, error);
+  if (!ret) {
+    g_debug ("DBus error");
+    return NULL;
+  }
+
+  g_variant_get (ret, "(v)", &inner);
+
+  return g_steal_pointer (&inner);
+}
+
```

Comment:
We should add a function for `ReadAll` too.

Reply:
aswell

Reply:
done. with tests